book/visual-effects.md

```diff
@@ -519,11 +551,236 @@ high-density enough to retire these techniques.
+- `raster_chrome` will draw the browser chrome to the `chrome_surface`.
+
+[^why-two-steps]: It might seem wasteful to copy from the chrome and
+    tab surface to an intermediate Skia surface, instead of directly
+    to the SDL surface. It is, but skipping that copy requires a lot
+    of tricky low-level code. In [Chapter 13](animations.md) we'll
+    avoid this copy in a different, better way.
+
+Let's start by doing the split:
+
+``` {.python}
+class Browser:
+    def raster_tab(self):
+        canvas = self.tab_surface.getCanvas()
+        canvas.clear(skia.ColorWHITE)
+        # ...
+
+    def raster_chrome(self):
+        canvas = self.chrome_surface.getCanvas()
+        canvas.clear(skia.ColorWHITE)
+        # ...
+
+    def draw(self):
+        canvas = self.root_surface.getCanvas()
+        canvas.clear(skia.ColorWHITE)
+        # ...
+```
+
+Since we didn't create the `tab_surface` on startup, we need to create
+it at the top of `raster_tab`:[^really-big-surface]
+
+[^really-big-surface]: For a very big web page, the `tab_surface` can
+be much larger than the size of the SDL window, and therefore take up
+a very large amount of memory. We'll ignore that, but a real browser
+would only paint and raster surface content up to a certain distance
+from the visible region, and re-paint/raster as the user scrolls.
+
+``` {.python}
+import math
+
+class Browser:
+    def raster_tab(self):
+        tab_height = math.ceil(self.active_tab.document.height + 2*VSTEP)
+
+        if not self.tab_surface or \
+                tab_height != self.tab_surface.height():
+            self.tab_surface = skia.Surface(WIDTH, tab_height)
+
+        # ...
+```
+
+The way we compute the page bounds here, based on the layout tree's
+height, would be incorrect if page elements could stick out below (or
+to the right) of their parents---but our browser doesn't support any
+features like that. Note that we need to recreate the tab surface if
+the page's height changes.
+
+Next, we need new code in `draw` to copy from the chrome and tab
+surfaces to the root surface. Moreover, we need to translate the
+`tab_surface` down by `chrome_bottom` and up by `scroll`, and clips it to
+only the area of the window that doesn't overlap the browser chrome:
+
+``` {.python}
+class Browser:
+    def draw(self):
+        # ...
+        
+        tab_rect = skia.Rect.MakeLTRB(
+            0, self.chrome.bottom, WIDTH, HEIGHT)
+        tab_offset = self.chrome.bottom - self.active_tab.scroll
+        canvas.save()
+        canvas.clipRect(tab_rect)
+        canvas.translate(0, tab_offset)
+        self.tab_surface.draw(canvas, 0, 0)
+        canvas.restore()
+
+        chrome_rect = skia.Rect.MakeLTRB(
+            0, 0, WIDTH, self.chrome.bottom)
+        canvas.save()
+        canvas.clipRect(chrome_rect)
+        self.chrome_surface.draw(canvas, 0, 0)
+        canvas.restore()
+
+        # ...
+```
+
+Note the `draw` calls: we are drawing the `tab_surface` and
```

Comment:
Some of this is redundant with the previous paragraph.

Reply:
Shortened.